core/message.go

```diff
@@ -93,19 +92,37 @@ func (msg *Message) GetPrevStreamID() MessageStreamID {
 
 // GetRouter returns the stream object behind the current StreamID.
 func (msg *Message) GetRouter() Router {
-	return StreamRegistry.GetRouterOrFallback(msg.GetStreamID())
+	return StreamRegistry.GetRouterOrFallback(msg.streamID)
 }
 
 // GetPrevRouter returns the stream object behind the previous StreamID.
 func (msg *Message) GetPrevRouter() Router {
 	return StreamRegistry.GetRouterOrFallback(msg.prevStreamID)
 }
 
+// GetOrigRouter returns the stream object behind the original StreamID.
+func (msg *Message) GetOrigRouter() Router {
+	return StreamRegistry.GetRouterOrFallback(msg.origStreamID)
+}
+
 // SetStreamID sets a new stream and stores the current one in the previous
-// stream field.
+// stream field. If the previous stream is invalid and the new stream is not,
+// the new stream will become the original stream.
 func (msg *Message) SetStreamID(streamID MessageStreamID) {
-	msg.prevStreamID = msg.GetStreamID()
-	msg.data.streamID = streamID
+	if streamID != InvalidStreamID && msg.streamID == InvalidStreamID {
+		msg.origStreamID = streamID
+	}
+	msg.prevStreamID = msg.streamID
+	msg.streamID = streamID
+}
+
+// SetOriginalStreamID acts like SetStreamID but always sets the original stream
+// ID, too. This method should be used before a message is routed for the first
+// time (e.g. in a consumer) or when the original stream should change.
+func (msg *Message) SetOriginalStreamID(streamID MessageStreamID) {
```

Comment:
SetlStreamIDAsOriginal is propably a better name as it extend SetStreamID by also setting the original